pkg/master/master.go

```diff
@@ -75,248 +69,36 @@ import (
-	ServiceReadWriteIP net.IP
-
-	// The range of ports to be assigned to services with type=NodePort or greater
-	ServiceNodePortRange util.PortRange
-
-	// Used to customize default proxy dial/tls options
-	ProxyDialer          apiserver.ProxyDialerFunc
-	ProxyTLSClientConfig *tls.Config
+	*genericapiserver.Config
 
+	EnableCoreControllers bool
+	EventTTL              time.Duration
+	KubeletClient         kubeletclient.KubeletClient
 	// Used to start and monitor tunneling
 	Tunneler Tunneler
-
-	// Additional ports to be exposed on the master service
-	// extraServicePorts is injectable in the event that more ports
-	// (other than the default 443/tcp) are exposed on the master
-	// and those ports need to be load balanced by the master
-	// service because this pkg is linked by out-of-tree projects
-	// like openshift which want to use the master but also do
-	// more stuff.
-	ExtraServicePorts []api.ServicePort
-	// Additional ports to be exposed on the master endpoints
-	// Port names should align with ports defined in ExtraServicePorts
-	ExtraEndpointPorts []api.EndpointPort
-
-	KubernetesServiceNodePort int
-}
-
-func (c *Config) storageDecorator() generic.StorageDecorator {
-	if c.EnableWatchCache {
-		return genericetcd.StorageWithCacher
-	}
-	return generic.UndecoratedStorage
 }
 
-type InstallSSHKey func(user string, data []byte) error
-
 // Master contains state for a Kubernetes cluster master/api server.
 type Master struct {
-	// "Inputs", Copied from Config
-	serviceClusterIPRange *net.IPNet
-	serviceNodePortRange  util.PortRange
-	cacheTimeout          time.Duration
-	minRequestTimeout     time.Duration
-
-	mux                      apiserver.Mux
-	muxHelper                *apiserver.MuxHelper
-	handlerContainer         *restful.Container
-	rootWebService           *restful.WebService
-	enableCoreControllers    bool
-	enableLogsSupport        bool
-	enableUISupport          bool
-	enableSwaggerSupport     bool
-	enableProfiling          bool
-	enableWatchCache         bool
-	apiPrefix                string
-	apiGroupPrefix           string
-	corsAllowedOriginList    []string
-	authenticator            authenticator.Request
-	authorizer               authorizer.Authorizer
-	admissionControl         admission.Interface
-	masterCount              int
-	apiGroupVersionOverrides map[string]APIGroupVersionOverride
-	requestContextMapper     api.RequestContextMapper
-
-	// External host is the name that should be used in external (public internet) URLs for this master
-	externalHost string
-	// clusterIP is the IP address of the master within the cluster.
-	clusterIP            net.IP
-	publicReadWritePort  int
-	serviceReadWriteIP   net.IP
-	serviceReadWritePort int
-	masterServices       *util.Runner
-	extraServicePorts    []api.ServicePort
-	extraEndpointPorts   []api.EndpointPort
-
-	// storage contains the RESTful endpoints exposed by this master
-	storage map[string]rest.Storage
+	*genericapiserver.Genericapiserver
```

Comment:
1. Embedding it made it easier to refactor.
2. Most of the public members and methods of generic api server should go away if we are able to move bulk of master.InstallAPI logic to generic apiserver (which I plan to do next).

That said, I think it makes sense to not embed genericapiserver here. Will see how big of a change that is. Is it fine to do that in a separate PR if its too big?